### ISSUE TITLE
twisted.spread.flavors.unjellyCached doesn't support new-style classes.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4447]
|<img alt="sirgolan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| sirgolan reported|
|-|-|
|Trac ID|trac#4447|
|Type|defect|
|Created|2010-05-22 01:16:50Z|
|Branch|https://github.com/twisted/twisted/tree/pb-newstyle-cache-refresh-4447|

I got the following traceback after converting some of my pb.Cacheable/pb.RemoteCache classes to new-style:

```
        Traceback (most recent call last):
          File "f:\Projects\Twisted\trunk\twisted\spread\banana.py", line 148, in gotItem
            self.callExpressionReceived(item)
          File "f:\Projects\Twisted\trunk\twisted\spread\banana.py", line 111, in callExpressionReceived
            self.expressionReceived(obj)
          File "f:\Projects\Twisted\trunk\twisted\spread\pb.py", line 514, in expressionReceived
            method(*sexp[1:])
          File "f:\Projects\Twisted\trunk\twisted\spread\pb.py", line 828, in proto_cachemessage
            self._recvMessage(self.cachedLocallyAs, requestID, objectID, message, answerRequired, netArgs, netKw)
        --- <exception caught here> ---
          File "f:\Projects\Twisted\trunk\twisted\spread\pb.py", line 840, in _recvMessage
            netResult = object.remoteMessageReceived(self, message, netArgs, netKw)
          File "F:\Projects\MV3D\branches\441-persistaccoun\mv3d\net\pb.py", line 620, in remoteMessageReceived
            kw)
          File "f:\Projects\Twisted\trunk\twisted\spread\flavors.py", line 418,in remoteMessageReceived
            args = broker.unserialize(args)
          File "f:\Projects\Twisted\trunk\twisted\spread\pb.py", line 776, in unserialize
            return unjelly(sexp, self.security, None, self)
          File "f:\Projects\Twisted\trunk\twisted\spread\jelly.py",line 1133, in unjelly
            return _Unjellier(taster, persistentLoad,invoker).unjellyFull(sexp)

          File "f:\Projects\Twisted\trunk\twisted\spread\jelly.py", line 629, in unjellyFull
            o = self.unjelly(obj)
          File "f:\Projects\Twisted\trunk\twisted\spread\jelly.py", line 666, in unjelly
            ret = thunk(obj[1:])
          File "f:\Projects\Twisted\trunk\twisted\spread\jelly.py", line 780, in _unjelly_tuple
            if isinstance(self.unjellyInto(l, elem, lst[elem]), NotKnown):
          File "f:\Projects\Twisted\trunk\twisted\spread\jelly.py", line 744, in unjellyInto
            o = self.unjelly(jel)
          File "f:\Projects\Twisted\trunk\twisted\spread\jelly.py", line 653, in unjelly
            val = method(self, obj)
          File "f:\Projects\Twisted\trunk\twisted\spread\flavors.py", line 501, in unjellyCached
            cProxy.__class__ = cNotProxy.__class__
        exceptions.TypeError: __class__ must be set to a class
```

It turns out that twisted.spread.flavors.unjellyCached doesn't support new-style classes. The attached diff fixes the problem for me.

Attachments:

* [diff.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/305/305335ad96b222a566cf71d88813765936270e67/cdf40f56efb0b30dc346ba10fa0d158ab4d25dde.txt) (536 bytes) - added by sirgolan on 2010-05-22 01:17:08Z - 
* [diff_test_pb.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/305/305335ad96b222a566cf71d88813765936270e67/df04f9fd1f84c563f2c793ce25d2229065ab6525.txt) (6378 bytes) - added by sirgolan on 2010-06-02 00:14:35Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4447 4447
type__defect defect
reporter__sirgolan sirgolan
priority__normal normal
milestone__ 
branch__branches_pb_newstyle_cache_refresh_4447 branches/pb-newstyle-cache-refresh-4447
branch_author__exarkun__sirgolan exarkun, sirgolan
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__ 
time__1274491010000000 1274491010000000
changetime__1308880641000000 1308880641000000
version__None None
owner__exarkun exarkun
cc__whitelynx cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to sirgolan|
|-|-|

Heya.  Thanks for the patch.  I wonder if you can distill the failure into a unit test as well?  I started to try to do that, but then I noticed `twisted/test/test_pb.py`, `NewStyleCachedTestCase`, which already tests basically the issue I thought you were describing.  So I'm not sure which case is actually broken here.

**Comment by twisted-trac** [comment]
|<img alt="sirgolan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|sirgolan set status to `assigned`|
|-|-|

I just threw a breakpoint in unjellyCached and ran NewStyleCachedTestCase. The breakpoint wasn't hit. I'm going to have a look at why and see if I can get a test to exercise this (or modify that test if it seems appropriate).

**Comment by twisted-trac** [comment]
|<img alt="sirgolan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|sirgolan commented|
|-|-|

I'm still not sure why NewStyleCachedTestCase isn't hitting that code, but I was able to basically clone BrokerTestCase.test_cache and use new style classes to expose the issue. It's probably still better to fix the other test if this is the case it intends to exercise.

**Comment by twisted-trac** [comment]
|<img alt="whitelynx's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|whitelynx commented|
|-|-|

Adding 'review' keyword so this can get looked at.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [31786]) Branching to 'pb-newstyle-cached-4447'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [31787]) Apply diff.txt and diff_test_pb.txt

refs [#4447](https://github.com/twisted/twisted/issues/4447)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

It looks like the existing support (and test) for this only covers the _initial_ transmission of the cacheable to the receiver.  On future sends of the cacheable is transmitted differently and it's that update-the-cacheable code path that doesn't handle new-style objects.

Understanding that, I think I can write a much simpler unit test for this case.  Which I will now attempt.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [32087]) Branching to 'pb-newstyle-cache-refresh-4447'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [32088]) Expand the existing new style cacheable test to exercise the re-send codepath

refs [#4447](https://github.com/twisted/twisted/issues/4447)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [32089]) Apply the original fix from sirgolan

refs [#4447](https://github.com/twisted/twisted/issues/4447)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun removed owner|
|-|-|

[Build results are here](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/pb-newstyle-cache-refresh-4447).  The fix is the original supplied by sirgolan, but I refactored it into a helper since there's another place in flavors.py that does the same thing.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_14"></a>|@jesstess set owner to @exarkun|
|-|-|

In ```twisted/test/test_pb.py```:

```
+                # Save a reference to it stays alive for the rest of this test
```

is missing a word. The docstring on the test also has a subject/verb agreement issue.

Other than that, looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [32158]) address review feedback

  * give the `test_newStyleCache` test method a better docstring
  * fix a typo in a comment in the test

refs [#4447](https://github.com/twisted/twisted/issues/4447)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set status to `closed`|
|-|-|

(In [32160]) Merge pb-newstyle-cache-refresh-4447

Author: sirgolan, exarkun
Reviewer: exarkun, jesstess
Fixes: [#4447](https://github.com/twisted/twisted/issues/4447)

Add support for new-style `RemoteCache` types to `twisted.spread`.
